@nx/js:tsc - wrong extends path used when generated temporary tsconfig

### Current Behavior


#### Background

When building with the `tsc` executor, a temporary tsconfig file gets generated for you in `{workspaceRoot}/tmp/{relativeProjectRoot}`

This temporary tsconfig file is meant to use the tsconfig `extends` syntax to extend the tsconfig in your project root with some custom behavior based on your build settings

ex:
```json
{
  "extends": "../../../../../packages/backend/db/tsconfig.lib.json"
}
```

#### The code

Internally, this executor is called [here](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/executors/tsc/tsc.impl.ts#L62)

The first thing it does is take the tsconfig path in your executor's configuration and [normalizes it](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/executors/tsc/tsc.impl.ts#L68). For example

```json
"build": {
  "executor": "@nx/js:tsc",
  "outputs": [
	"{options.outputPath}"
  ],
  "options": {
	"outputPath": "dist/packages/shared/data-types",
	"main": "packages/shared/data-types/src/index.ts",
	"tsConfig": "packages/shared/data-types/tsconfig.lib.json"
  }
}
```

will turn the relative `tsConfig` path into `'/home/user/Documents/git/my-project/packages/backend/db/tsconfig.lib.json'`

This new absolute path is then passed down to create the temporary tsconfig file in the following steps
1. This line to call into `checkDependencies` [here](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/executors/tsc/tsc.impl.ts#L70-L73)
2. `checkDependencies` calls `createTmpTsConfig` [here](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/utils/check-dependencies.ts#L29-L34)
3. The exact value for the `extends` field is calculated in `readTsConfigWithRemappedPaths` [here](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/utils/buildable-libs-utils.ts#L443-L447)
4. And ``readTsConfigWithRemappedPaths` then calculated the exact path [here](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/utils/buildable-libs-utils.ts#L203-L206)

#### The bug

Note that for the value of `extends` to be calculated correctly in [readTsConfigWithRemappedPaths](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/utils/buildable-libs-utils.ts#L203-L206), you have two options:
1. Make both paths (`dirnameTsConfig`, `tsConfig`) absolute
2. Make both paths relative to `workspaceRoot`

However, instead, right now `dirnameTsConfig` is turned into a relative path while `tsConfig` is kept as an absolute path (remember we normalized it [here](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/js/src/executors/tsc/tsc.impl.ts#L68))

This results in the value of `generatedTsConfig.extends` to be incorrect, causing builds to fail

### Expected Behavior

The paths should either both be absolute, or both be relative.

The easiest way to do this was to pass in the relative path into the `checkDependencies` function (since the `tsConfig` variable is not accessed anywhere except to calculate `generatedTsConfig.extends`), but changing `dirnameTsConfig` to always be absolute would also be okay

### GitHub Repo

https://github.com/SebastienGllmt/nx-tsc-repro

### Steps to Reproduce

See repro README

#### Builds seems to work fine for me on my repo.... how come you can reproduce it?

The reason this error wasn't caught before is presumably because if the `extends` path points to an invalid path, builds will still work if they were going to work with the default tsc configuration.

Notably, if you add `console.log(tsConfig)`  [here](https://github.com/nrwl/nx/blob/c2af06e0a12529a5c3d1a09024ccc1fabf4e5604/packages/workspace/src/utilities/typescript/compilation.ts#L145), you can see that these builds actually result in the following error that gets supressed

```json5
errors: [
    {
      file: undefined,
      start: undefined,
      length: undefined,
      messageText: "Cannot read file '/home/user/Documents/git/my-project/tsconfig.lib.json'.",
      category: 1,
      code: 5083,
      reportsUnnecessary: undefined,
      reportsDeprecated: undefined
    },
    {
      file: undefined,
      start: undefined,
      length: undefined,
      messageText: `No inputs were found in config file 'tsconfig.json'. Specified 'include' paths were '["**/*"]' and 'exclude' paths were '[]'.`,
      category: 1,
      code: 18003,
      reportsUnnecessary: undefined,
      reportsDeprecated: undefined
    }
  ],
```

#### Reliably causing an actual build failure

Although these builds always have a configuration error, as I mentioned that doesn't always actually lead to a build error
 
2. `nx build`

### Nx Report

```
Node           : 20.11.1
OS             : linux-x64
Native Target  : x86_64-linux
npm            : 10.2.4

nx                 : 19.8.2
@nx/js             : 19.8.2
@nx/linter         : 19.8.2
@nx/eslint         : 19.8.2
@nx/workspace      : 19.8.2
@nx/devkit         : 19.8.2
@nx/eslint-plugin  : 19.8.2
@nrwl/tao          : 19.8.2
@nx/vite           : 19.8.2
@nx/web            : 19.8.2
typescript         : 5.6.2
---------------------------------------
Registered Plugins:
@nx/eslint/plugin
```

### Failure Logs

```
error TS5090: Non-relative paths are not allowed when 'baseUrl' is not set. Did you forget a leading './'?

Cannot read file '/home/user/Documents/git/my-project/tsconfig.lib.json'
```

### Package Manager Version

pnpm 9.6.0

### Operating System

- [ ] macOS
- [x] Linux
- [ ] Windows
- [ ] Other (Please specify)

### Additional Information

_No response_

--- a/packages/js/src/utils/buildable-libs-utils.spec.ts
+++ b/packages/js/src/utils/buildable-libs-utils.spec.ts
@@ -1,10 +1,14 @@
 import { DependencyType, ProjectGraph, TaskGraph } from '@nx/devkit';
+import { TempFs } from '@nx/devkit/internal-testing-utils';
+import { readFileSync } from 'fs';
 import {
-  calculateProjectDependencies,
   calculateDependenciesFromTaskGraph,
+  calculateProjectDependencies,
+  createTmpTsConfig,
   DependentBuildableProjectNode,
   updatePaths,
 } from './buildable-libs-utils';
+import { join } from 'path';
 
 describe('updatePaths', () => {
   const deps: DependentBuildableProjectNode[] = [
@@ -766,5 +770,47 @@ describe('missingDependencies', () => {
     expect(() =>
       calculateProjectDependencies(graph, 'root', 'example', 'build', undefined)
     ).toThrow();
+  });
+});
+
+describe('createTmpTsConfig', () => {
+  it('should create a temporary tsconfig file extending the provided tsconfig', () => {
+    const fs = new TempFs('buildable-libs-utils#createTmpTsConfig');
+    const tsconfigPath = 'packages/foo/tsconfig.json';
+    fs.createFileSync(tsconfigPath, '{}');
+
+    const tmpTsConfigPath = createTmpTsConfig(
+      tsconfigPath,
+      fs.tempDir,
+      'packages/foo',
+      []
+    );
+
+    const tmpTsConfig = readFileSync(tmpTsConfigPath, 'utf8');
+    // would be generated at <workspaceRoot>/tmp/packages/foo/build/tsconfig.generated.json
+    // while the extended tsconfig path is <workspaceRoot>/packages/foo/tsconfig.json
+    expect(JSON.parse(tmpTsConfig).extends).toBe(
+      '../../../../packages/foo/tsconfig.json'
+    );
+  });
+
+  it('should also work when the provided tsconfig is an absolute path', () => {
+    const fs = new TempFs('buildable-libs-utils#createTmpTsConfig');
+    const tsconfigPath = join(fs.tempDir, 'packages/foo/tsconfig.json');
+    fs.createFileSync(tsconfigPath, '{}');
+
+    const tmpTsConfigPath = createTmpTsConfig(
+      tsconfigPath,
+      fs.tempDir,
+      'packages/foo',
+      []
+    );
+
+    const tmpTsConfig = readFileSync(tmpTsConfigPath, 'utf8');
+    // would be generated at <workspaceRoot>/tmp/packages/foo/build/tsconfig.generated.json
+    // while the extended tsconfig path is <workspaceRoot>/packages/foo/tsconfig.json
+    expect(JSON.parse(tmpTsConfig).extends).toBe(
+      '../../../../packages/foo/tsconfig.json'
+    );
   });
 });

--- a/packages/js/src/utils/buildable-libs-utils.ts
+++ b/packages/js/src/utils/buildable-libs-utils.ts
@@ -16,7 +16,7 @@ import { unlinkSync } from 'fs';
 import { isNpmProject } from 'nx/src/project-graph/operators';
 import { directoryExists, fileExists } from 'nx/src/utils/fileutils';
 import { output } from 'nx/src/utils/output';
-import { dirname, join, relative, isAbsolute, extname } from 'path';
+import { dirname, join, relative, extname, resolve } from 'path';
 import type * as ts from 'typescript';
 import { readTsConfigPaths } from './typescript/ts-config';
 
@@ -194,18 +194,23 @@ function collectDependencies(
 }
 
 function readTsConfigWithRemappedPaths(
-  tsConfig: string,
-  generatedTsConfigPath: string,
-  dependencies: DependentBuildableProjectNode[]
+  originalTsconfigPath: string,
+  generatedTsconfigPath: string,
+  dependencies: DependentBuildableProjectNode[],
+  workspaceRoot: string
 ) {
   const generatedTsConfig: any = { compilerOptions: {} };
-  const dirnameTsConfig = dirname(generatedTsConfigPath);
-  const relativeTsconfig = isAbsolute(dirnameTsConfig)
-    ? relative(workspaceRoot, dirnameTsConfig)
-    : dirnameTsConfig;
-  generatedTsConfig.extends = relative(relativeTsconfig, tsConfig);
+  const normalizedTsConfig = resolve(workspaceRoot, originalTsconfigPath);
+  const normalizedGeneratedTsConfigDir = resolve(
+    workspaceRoot,
+    dirname(generatedTsconfigPath)
+  );
+  generatedTsConfig.extends = relative(
+    normalizedGeneratedTsConfigDir,
+    normalizedTsConfig
+  );
   generatedTsConfig.compilerOptions.paths = computeCompilerOptionsPaths(
-    tsConfig,
+    originalTsconfigPath,
     dependencies
   );
 
@@ -443,7 +448,8 @@ export function createTmpTsConfig(
   const parsedTSConfig = readTsConfigWithRemappedPaths(
     tsconfigPath,
     tmpTsConfigPath,
-    dependencies
+    dependencies,
+    workspaceRoot
   );
   process.on('exit', () => cleanupTmpTsConfigFile(tmpTsConfigPath));
 

